teps/0033-tekton-feature-gates.md

```diff
@@ -0,0 +1,153 @@
+  on the current opt-out behavior.
+- "Tekton Bundles" were introduced in the `Pipeline` API, which has a *beta*
+  level of maturity. Bundles however are experimental, and we have no mechanism
+  in place to express the *alpha* level of maturity for a feature that is
+  controlled via new fields in a *beta* API.
+
+To solve the case of Tekton Bundles, as well as that of custom tasks, we used a
+feature flag and introduced the features as disabled by default.
+
+
+### Goals
+
+- Define and document what is the recommended strategy for different kind
+  of changes to Tekton (e.g. backward compatible vs. incompatible,
+  addition to the API, changes to existing APIs, removals and possible more)
+- Document the interface to the user, how do we signal changes and features,
+  and their level of maturity (e.g. documentation, release notes, config maps,
+  controller flag, api versions)
+- Document the process for introducing a backward compatibile feature and
+  its lifecycle: which are the levels of maturity, which documents have
+  to be updated, what goes into release notes, how does a feature move
+  to a different level of maturity, what is the test strategy
+- Document the process for introducing a backward incompatibile change and
+  its lifecycle
+- Identify tools required to develop, test and document combination of features
+  and API versions. Examples:
+  - tools that help adding a feature flag or documentation on the process
+  - tools to implement feature matrix testing
+  - automation or recommendation on documentation to be produced
+  - tools to add a new API version
+- Identify projects that are impacted. Examples:
+    - UIs (cli, dashboard): support for experimental features and API versions
+    - Catalog and Hub: how to version tasks that rely on experimental features
+    - Operator: how to control which versions and features are deployed / enabled
+
+
+### Non-Goals
+
+- Impacted projects should used dedicated TEPs to design their handling of
+  experimental featurs and flags
+
+### Use Cases
+
+- As a Tekton developer I want to introduce changes to Tekton in a format
+  such as users do not rely on them until such changes are proven and ready
+  to be made generally available
+- As a Tekton developer I want to introduce an experimental feature as part
+  of a beta or stable API
+- As a Tekton develoloper I want to introduce an experimental behaviour in
+  Tekton as opt-in
+
+## Requirements
+
+- It should be possible to introduce features and API changes as opt-in
+- It should be possible to transition features and API changes to become
+  opt-out once they are deemed stable enough
+- Document available feature flags, how to use them, when they were
+  introduced the level of maturity of the associated feature
+- Tools to test enabled/disabled feature flags as part of the test matrix
+- It should be possible for a Tekton operator to restrict access to
+  feature flags (*not* on a one-by-one case)
```

Comment:
i think the rest of this doc does a great job of not trying to specify a solution - but these last 3 requirements seem to assume we're going with feature flags. maybe we can reword them as something like:

1. The maturity of each feature should be clearly documented and it should be clear how to opt into optional features
2. Tools should be available to test all desired combinations of features at various maturity levels
3. It should be possible for a Tekton operator to restrict access to alpha features

I'm not 100% sure about (3): if we go with feature flags this is possible, but i dont totally understand if we actually _need_ it - do we want to be able to say that Tekton operators can restrict access to all alpha features, e.g. even alpha CRDS? (if yes then (3) makes sense to me)